_posts/2022-03-15-turn-based-offline-rl.md

```diff
@@ -0,0 +1,344 @@
+out of the N runs, we will pick the one that gave us the best results (in terms of average reward).
+However, please note that this may be too optimistic and could lead to unexpected behavior on production.
+A better approach would be actually the one that takes into account the “average policy”. But, to “average out”
+policies is not a trivial task. We discuss this aspect in detail in the final section.
+
+## Results
+
+### Visualizing the agent “in turns”
+
+First, we ask ourselves the following question: does the agent learn “in turns”? We can check this
+by visualizing subsequent state-action visitation maps:
+
+![Figure 5](/img/articles/2022-03-15-turn-based-offline-rl/agent_learns_in_turns_reduced.png)
+*Fig.5 — Visualization of subsequent state-action visitation heatmaps. Here we visualize 4 subsequent turns — after one
+single turn the agent learns the fastest path to reach the target cell. As long as we proceed in turns, the agent
+improves its performance, eventually approaching a behaviour comparable to the optimal policy.*
+
+Our algorithm seems to work! When starting with a uniform policy, we can see that the agent even after a
+single turn the agent quickly learns the fastest path to reach the target cell. As long as we proceed in turns, the model will
+consistently improve its performance, by quickly getting to the “R” cell even more often. In this sense, visitation
+maps get closer to the optimal one where the agent basically reaches the target in a few steps and then just stays
+there, without performing any further steps.
+
+### Does the agent improve its performance over time?
+
+How many turns are needed to start having results comparable to
+the optimal policy? In other words, how much better are we performing if compared to not doing any turn at all?
+
+Let’s analyze the plot below. In this figure we are plotting the algorithm’s performance, measured in
+"averaged reward" (the higher, the better), as the amount of data available offline increases. In general, we expect
+the curve obtained by running the optimal policy (violet curve) to represent an upper bound: it’s the best we
+can achieve! On the other hand, we expect the curve obtained by running the random policy without any "turn"
+(green curve), to be our lower bound. Also, generally speaking, we expect that the more offline data is available,
+the better the achieved scores will be, since our chosen Offline RL will have more data coverage and possibility
+to converge to the optimal policy. Given this, we can observe that the performance of the turn-based procedure falls
+in the middle between the aforementioned upper and lower bounds: as the number of turns increases, the closer we get
+to the upper bound. However, we can observe that 3 turns are already enough to start having better performance than
+the lower bound. This plot confirms our hypothesis: "Turn-based Offline RL" stands exactly between Online RL
```

Comment:
```suggestion
the lower bound. This plot confirms our hypothesis: “Turn-based Offline RL“ stands exactly between Online RL
```

Reply:
Fixed (and I checked the rest of the text: no more of such mistakes!)